### PR TITLE
[#394] Provide non-reflective access to JVM Attach API on JDK9+

### PR DESCRIPTION
**Pull request by grgrzybek**

### CODE REVIEW

**Comment by grgrzybek** [comment]
Please do not merge yet. Work in progress.

File: agent/jvm/pom.xml [hunk 1]
```diff
@@ -96,16 +96,15 @@
       <plugin>
         <artifactId>maven-compiler-plugin</artifactId>
         <configuration>
-          <source>1.6</source>
-          <target>1.6</target>
+          <source>1.7</source>
+          <target>1.7</target>
```
Comment (rhuss) [inline agent/jvm/pom.xml:100]: I guess this jump is required, and I'm not against it. But we need to bump the version number accordingly, to indicate that the release is not compatible with Java 1.6. anymore. Also, I'd prefer to have this as a global switch for all components (not only for the JVM agent)

File: agent/jvm/src/main/java/org/jolokia/jvmagent/security/asn1/DERInteger.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jolokia.jvmagent.security.asn1;
+
+public class DERInteger implements DERObject {
```
Comment (rhuss) [inline agent/jvm/src/main/java/org/jolokia/jvmagent/security/asn1/DERInteger.java:19]: Are those ASN1 objects copied over from some other open-source library (like bouncycastle) or do you have written those ? 

It's probably over the top here, but are there any tests for those classes ?

**Review by rhuss — COMMENTED** [review]
Thanks a lot for all of your efforts, looks good to me in general ! Maybe it would have been better to split it up in two PRs (one for the reflection in the VirtualMachineHandler, one for the CERT handling).

I have two questions though, see the comments inline. Also if the ASN.1 objects are borrowed from somewhere else we should check the license and add the corresponding license headers/files. For the switch to 1.7 I wonder whether this is required, and if so, we should do it at the top-level.

File: agent/jvm/pom.xml (hunk 1 above)
Comment (grgrzybek) [reply agent/jvm/pom.xml:100]: It is required, when I tried to run the tests using JDK17 (as it no longer can handle `target=1.6`), but it's not a requirement for any other change. I can skip this change.

File: agent/jvm/src/main/java/org/jolokia/jvmagent/security/asn1/DERInteger.java (hunk 2 above)
Comment (grgrzybek) [reply agent/jvm/src/main/java/org/jolokia/jvmagent/security/asn1/DERInteger.java:19]: The tests are in `org/jolokia/jvmagent/security/asn1/DEREncodingTest.java` and I confirm - I looked a bit at BC classes (only to check what _DER Tag_ is), but every line of code is written by me. I was consulting only the X.680/X.690 specifications.

**Comment by grgrzybek** [comment]
First - this PR can be treated as "for review only" and I agree that separate PRs are better.

DER classes are written by me, as I didn't want to borrow anything from BC (or internal JDK classes). I also didn't want to introduce any dependency to external, lightweight ASN.1/DER libraries.
There are few missing DER/ASN.1 constructs, because I needed only the ones for X.509 certificate structure.

**Comment by grgrzybek** [comment]
I'm just checking the final part - `@JsonMBean` annotation handling. When I'm ready, I'll update this PR (force pushing probably), but eventually I'll split this PR into 2 (or 3) separate ones.

**Comment by rhuss** [comment]
> DER classes are written by me, as I didn't want to borrow anything from BC (or internal JDK classes). I also didn't want to introduce any dependency to external, lightweight ASN.1/DER libraries.

ah, very cool. Do you think we should some unit tests for those ? They don't look trivial

**Comment by rhuss** [comment]
I think we can keep it 1 PR, don't bother with fiddling around with multiple PRs. Maybe you can have several commits per feature, but happy also if not. Don't waste efforts here.

**Comment by grgrzybek** [comment]
> ah, very cool. Do you think we should some unit tests for those ? They don't look trivial

the tests are in `org/jolokia/jvmagent/security/asn1/DEREncodingTest` ;)

**Comment by grgrzybek** [comment]
And #466 is a separate PR for cert generation.

File: agent/jvm/pom.xml (hunk 1 above)
Comment (rhuss) [reply agent/jvm/pom.xml]: Maybe we can skip it for now, and consider to bump to 1.7 as minimum in a followup PR (and then already on the overall level) ?

Comment (grgrzybek) [reply agent/jvm/pom.xml]: I reverted this change, so it's still 1.6. 1.7 was needed if I wanted to build using JDK17. But I needed it to _test_ using JDK17. After verifying it works, we can leave at 1.6 value ;)

**Comment by rhuss** [comment]
Sorry that I overlooked the tests ;( going to continue tomorrow on the PRs ...

**Review by rhuss — APPROVED** [review]
Looks good to me !

**Pull request merged**